.github/workflows/openstack/prune_openstack

```diff
@@ -1,84 +1,83 @@
-}
+my $time_cmd = qq/ date -d '-2 hour' --utc +"%Y-%m-%dT%H:%M:%SZ"/;
 
-remove_stale_instances () {
-    echo "HAMMERTIME: $HAMMER_TIME"
-    while IFS="," read -r ID CREATED_DATE NAME
-    do
-        if [[ $CREATED_DATE < $HAMMER_TIME ]];
-        then
-            echo "deleting: $ID,$CREATED_DATE,$NAME"
-            openstack server delete "$ID"
-        fi
-    done < <(cat id_created_name.txt)
-}
+my $hammer_time = `$time_cmd`;
 
-get_keys () {
-    LIST=$(openstack keypair list -f json  | jq -r .[].Name)
-    rm keyid_created_name.txt > /dev/null 2>&1
-    while IFS= read -r KEYPAIR; do
-        openstack keypair show -f json "$KEYPAIR" | jq -r '.id + "," + .created_at + "," + .name' | grep $KEY_NAME >> keyid_created_name.txt
-    done <<< "$LIST"
-}
+remove_stale_instances();
 
-remove_stale_keys () {
-    echo "HAMMERTIME: $HAMMER_TIME"
-    while IFS="," read -r ID CREATED_DATE NAME
-    do
-        if [[ $CREATED_DATE < $HAMMER_TIME ]];
-        then
-            echo "deleting: $ID,$CREATED_DATE,$NAME"
-            openstack keypair delete "$ID"
-        fi
-    done < <(cat keyid_created_name.txt)
-}
+remove_stale_keys();
+
+sub get_keys {
+    my $cmd = qq{ openstack keypair list -f json  | jq -r .[].Name };
+
+    my $keys = {};
+
+    my @list = split( '\n', eval { `$cmd` } );
+    die $@ if $@;
+
+    foreach my $key_name (@list) {
+        my $created_cmd = qq{ openstack keypair show -f json "$key_name" | jq -r '.id + "," + .created_at + "," + .name' };
+        ( $keys->{$key_name}->{'id'}, $keys->{$key_name}->{'created_on'}, $keys->{$key_name}->{'name'} ) = split( ',', eval { `$created_cmd` } );
+    }
 
-get_volumes () {
-    rm volume_ids.txt > /dev/null 2>&1
-    openstack volume list | grep packer  | grep available | awk '{print $2}' >> volume_ids.txt
+    return $keys;
 }
 
-remove_volumes () {
-    while IFS= read -r ID; do
-        openstack volume delete "$ID" || true
-    done < <(cat volume_ids.txt)
+sub get_instances {
+    my $cmd = qq{ openstack server list -f json --no-name-lookup | jq -r .[].ID };
+
+    my $instances = {};
+
+    my @list = split( '\n', eval { `$cmd` } );
+    die $@ if $@;
+
+    foreach my $VM (@list) {
+        my $created_cmd = qq{ openstack server show -f json "$VM" | jq -r '.id + "," + .created + "," + .name' };
+        ( $instances->{$VM}->{'id'}, $instances->{$VM}->{'created_on'}, $instances->{$VM}->{'name'} ) = split( ',', eval { `$created_cmd` } );
+    }
+
+    return $instances;
 }
 
-cleanup () {
-    rm -fv id_created_name.txt
-    rm -fv keyid_created_name.txt
+sub remove_stale_keys {
+    my $keys = get_keys();
+
+    my @list = grep { $_->{'name'} =~ /${\KEY_NAME}/ } map { $instances->{$_} } keys %{$keys};
+
+    foreach my $key (@list) {
+        if ( $instances->{$id}->{'created_on'} < $hammer_time ) {
+            print "deleting: Key ID: $key->{'name'}, created_on: $key->{'created_on'}, hammer_time: $hammer_time\n";
+            system("openstack keypair delete $ID");
```

Comment:
Do you need to verify exit code here? I notice you're not using a precise path to access the `openstack` command. what $PATH are you assuming here?

Reply:
Hard-coded the path in for the ubuntu openstack-cli as that is where this is going to be running runner wise.